src/timeout.py

```diff
@@ -20,96 +20,164 @@
 from base import Service
 
 
-class Timeout(Service):
+class TimeoutService(Service):
 
-    def __init__(self, configs, args):
-        super().__init__(configs, args, 'timeout')
+    def __init__(self, configs, args, name):
+        super().__init__(configs, args, name)
         self._poll_period = args.poll_period
 
-    def _set_node_result(self, node):
-        """Set node result of timed out nodes
-        If the node is in 'running' state, set the node result to 'incomplete'.
-        When the timed out node is not in 'running' state, set the node result
-        to 'fail' if one of the child nodes failed. Otherwise, set the result
-        to 'pass'.
-        """
-        if node['state'] == 'running':
-            node['result'] = 'incomplete'
-            return node
-        failed_child_nodes = self._db.count_nodes({
-            'parent': node['_id'],
-            'result': 'fail'
-        })
-        node['result'] = 'fail' if failed_child_nodes else 'pass'
-        return node
-
-    def _update_available_node(self, node, current_time):
-        """Set node state to 'done' if holdoff is expired and all child nodes
-        are completed.
-        Set node state to `closing` if holdoff is expired and at least one of
-        the child nodes is incomplete"""
-        holdoff_time = datetime.fromisoformat(node['holdoff'])
-        if current_time > holdoff_time:
-            self._logger.log_message(logging.INFO,
-                                     f"Holdoff reached for {node['_id']}")
-            total_child_nodes = self._db.count_nodes({
-                'parent': node['_id'],
+    def _get_pending_nodes(self, filters=None):
+        nodes = {}
+        node_filters = filters.copy() if filters else {}
+        for state in ['running', 'available', 'closing']:
+            node_filters['state'] = state
+            for node in self._db.get_nodes(node_filters):
+                nodes[node['_id']] = node
+        return nodes
+
+    def _count_running_child_nodes(self, parent_id):
+        nodes_count = 0
+        for state in ['running', 'available', 'closing']:
+            nodes_count += self._db.count_nodes({
+                'parent': parent_id, 'state': state
             })
```

Comment:
I see you have used a different way of getting pending child nodes.
I used a simple way by calculating `all_child_nodes - completed_child_nodes`

Reply:
Yes, I know